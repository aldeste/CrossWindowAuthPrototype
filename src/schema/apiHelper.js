/* @flow */
import DataLoader from "dataloader";
import chalk from "chalk";
import { Person as dbPerson, Planet as dbPlanet } from "../data";
import User, { type Viewer } from "../models/User";
import Planet from "../models/Planet";

import { type GraphQLObjectType } from "graphql";

export type Loaders = {
  Person: Function,
  Planet: Function
};

export type DataLoaders = {
  Person: DataLoader<*, *>,
  Planet: DataLoader<*, *>
};

/**
 * This resolves a query to each id, batching the array into a
 * single fetch. this function is memoized as a response to DataLoader,
 * meaning the same ID won't be fetched twice in the same query.
 */
export function getResolve(
  type: Class<*>,
  ids: Array<string>,
  options: Object = {}
): Promise<*> {
  return Promise.all(
    ids.map(id => {
      // We console log each request to ilustrate how dataloader works,
      // If the same reqouse is requisted twice durning the same request
      // it'll still only consol log once, since the result is memoized.
      console.log(
        chalk.bold.magenta(
          "Fetched",
          chalk.yellow.italic(type.name),
          "with id",
          chalk.yellow.italic(id),
          "from database"
        )
      );
      return type.scope("withIds").findById(id, options);
    })
  );
}

/**
 * A factory to create DataLoader memoized database calls. Best
 * practice is to create each per request, this will help if
 * same data is fetched multiple times durning the same request.
 */
export function createLoaders(Viewer: ?Viewer): Loaders {
  const loaders: DataLoaders = {
    Person: new DataLoader(ids => getResolve(dbPerson, ids)),
    Planet: new DataLoader(ids => getResolve(dbPlanet, ids))
  };

  return {
    Person: (id: string): Promise<?User> =>
      User.gen(Viewer || null, id, loaders),
    PersonConnection: ids => User.genMany(Viewer || null, ids, loaders),
    Planet: (id: string): Promise<?Object> =>
      Planet.gen(Viewer || null, id, loaders)
  };
}

/**
 * Given a type and ID, get the object with the ID.
 */
export async function getObjectFromTypeAndId(
  type: Object,
  id: string,
  context?: Object
): Promise<Object> {
  const Loaders = (context && context.loaders) || createLoaders();
  return await Loaders[type.name](id);
}

/**
 * Given a type, fetch all objects
 */
export async function getObjectsByType(
  type: GraphQLObjectType,
  args?: ?Object,
  context: Object,
  ids: Array<string>
): Promise<Object> {
  const Loaders = (context && context.loaders) || createLoaders();
  const objects = await Loaders[type.name + "Connection"](ids);

  return { objects, currentCount: objects.length };
}
