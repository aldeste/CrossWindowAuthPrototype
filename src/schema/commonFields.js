// @flow

/**
 * This file contains fields used by each GraphQL type. Instead
 * of defining them multiple times, they are defined here.
 */

import { GraphQLString, type GraphQLFieldConfig } from "graphql";

export function createdField(): GraphQLFieldConfig<*, *> {
  return {
    type: GraphQLString,
    description:
      "The ISO 8601 date format of the time that this resource was created.",
    resolve: current => current.created
  };
}

export function editedField(): GraphQLFieldConfig<*, *> {
  return {
    type: GraphQLString,
    description:
      "The ISO 8601 date format of the time that this resource was edited.",
    resolve: current => current.edited
  };
}
