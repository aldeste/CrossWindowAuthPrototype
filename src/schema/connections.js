/* @flow */
import {
  connectionFromArray,
  connectionArgs,
  connectionDefinitions
} from "graphql-relay";
import { getObjectsByType } from "./apiHelper";
import {
  type GraphQLObjectType,
  type GraphQLFieldConfig,
  GraphQLInt
} from "graphql";

export function connectionFromType(
  name: string,
  type: GraphQLObjectType,
  idField: string
): GraphQLFieldConfig<*, *> {
  const { connectionType } = connectionDefinitions({
    name,
    nodeType: type,
    resolveNode: edge => edge.node,
    connectionFields: () => ({
      currentCount: {
        type: GraphQLInt,
        resolve: conn => conn.currentCount,
        description: `A count of the total number of objects in this connection, ignoring pagination.
This allows a client to fetch the first five objects by passing "5" as the
argument to "first", then fetch the total count so it could display "5 of 83",
for example.`
      }
    })
  });
  return {
    type: connectionType,
    args: connectionArgs,
    resolve: async (currentType: Object, args: Object, ctx: Object) => {
      const { objects, currentCount } = await getObjectsByType(
        type,
        args,
        ctx,
        currentType[idField].map(x => x.id)
      );
      return {
        ...connectionFromArray(objects, args),
        currentCount
      };
    }
  };
}
