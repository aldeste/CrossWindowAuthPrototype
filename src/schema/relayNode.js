// @flow
import { fromGlobalId, nodeDefinitions } from "graphql-relay";

import PersonType from "./types/personType";
import PlanetType from "./types/planetType";

import { getObjectFromTypeAndId } from "./apiHelper";

export function idFetcher(
  globalId: string,
  context: Object
): Promise<Object> | null {
  const { type, id } = fromGlobalId(globalId);
  const resolve = (type: Object) => getObjectFromTypeAndId(type, id, context);
  if (type === "people") {
    return resolve(PersonType);
  }
  if (type === "planets") {
    return resolve(PlanetType);
  }
  return null;
}

export function typeResolver(object: { GraphQLType?: Object }): Object | null {
  if (object.GraphQLType === PersonType.name) {
    return PersonType;
  }
  if (object.GraphQLType === PlanetType.name) {
    return PlanetType;
  }
  return null;
}

const { nodeInterface, nodeField } = nodeDefinitions(idFetcher, typeResolver);

export { nodeInterface, nodeField };
