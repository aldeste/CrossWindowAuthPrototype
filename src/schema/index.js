// @flow
import { GraphQLID, GraphQLObjectType, GraphQLSchema } from "graphql";
import { fromGlobalId } from "graphql-relay";

import { nodeField } from "./relayNode";
import personType from "./types/personType";
import { getObjectFromTypeAndId } from "./apiHelper";

type arguments = { personId?: ?string, id?: ?string };
type globalIdObejct = { type?: ?string, id?: ?string };

const Root = new GraphQLObjectType({
  name: "Root",
  fields: () => ({
    viewer: {
      type: personType,
      args: {
        id: { type: GraphQLID },
        personId: { type: GraphQLID }
      },
      resolve(_: *, args: arguments, viewer: Object): Object {
        const getter = (id: string) =>
          getObjectFromTypeAndId(personType, id, viewer);
        const { personId, id: globalId } = args;

        if (typeof personId === "string") {
          return getter(personId);
        }

        if (typeof globalId === "string") {
          const { id, type }: globalIdObejct = fromGlobalId(globalId);
          if (!id || id === "" || type === "") {
            throw new Error(`No valid ID extracted from ${globalId}`);
          }

          return getter(id);
        }
        throw new Error(`You must provide id or personId`);
      }
    },
    node: nodeField
  })
});

export default new GraphQLSchema({ query: Root });
