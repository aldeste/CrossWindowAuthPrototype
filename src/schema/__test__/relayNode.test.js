import { idFetcher, typeResolver } from "../relayNode";
import schema from "../";
import { graphql } from "graphql";
import { createLoaders } from "../apiHelper";

beforeEach(async () => {
  console.log = jest.fn();
});

describe("idFetcher", () => {
  it("should default to null if unresolvable", () => {
    expect(idFetcher("IiI6IiI=")).toBeNull();
  });

  it("should return an object if parameters are valid ", async () => {
    await require("../../data").initializeDatabase();
    const result = await idFetcher("UGVyc29uOjQ=");
    expect(result.name).toBeDefined();
    expect(result.token).toBeDefined();
    expect(result.id).toBeDefined();
  });
});

describe("typeResolver", () => {
  it("should default to null if unresolvable", () => {
    expect(typeResolver({})).toBeNull();
  });

  it("should return correct type", () => {
    const obj = {
      GraphQLType: "Person"
    };
    expect(typeResolver(obj).name).toBe("Person");
  });
});

describe("nodeInterface", () => {
  beforeEach(async () => {
    await require("../../data").initializeDatabase();
  });

  const rootValue = {};
  const context = { loaders: createLoaders() };

  it("should return field with id", async () => {
    const query = `query Test {
      node(id:"UGVyc29uOjQ=") {
        ... on Person {
          name
        }
      }
    }`;

    const { data } = await graphql(schema, query, rootValue, context);
    expect(data).toMatchSnapshot();
  });

  it("should error if no id is suplied", async () => {
    const query = `query Test {
      node {
        ... on Person {
          name
        }
      }
    }`;

    const result = await graphql(schema, query, rootValue, context);

    expect(result.errors).toBeDefined();
  });

  it("should fail if called with false value", async () => {
    const query = `query Test {
      node(id:"e") {
        ... on Person {
          name
        }
      }
    }`;

    const { data } = await graphql(schema, query, rootValue, context);
    expect(data).toMatchSnapshot();
  });
});
