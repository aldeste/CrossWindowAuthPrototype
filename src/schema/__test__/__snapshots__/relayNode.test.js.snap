// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodeInterface should fail if called with false value 1`] = `
Object {
  "node": null,
}
`;

exports[`nodeInterface should return field with id 1`] = `
Object {
  "node": Object {
    "name": "Darth Vader",
  },
}
`;
